packages/plugin-typescript/sources/index.ts

```diff
@@ -41,9 +50,27 @@ const afterWorkspaceDependencyAddition = async (
   if (nonNullSuggestions.length === 0)
     return;
 
-  const selected = nonNullSuggestions[0].descriptor;
-  if (selected === null)
-    return;
+  let selected: Descriptor | null | undefined = null;
+
+  // Try reusing an existing version
+  if (strategies.includes(suggestUtils.Strategy.REUSE))
+    selected = nonNullSuggestions.find(suggestion => suggestion.reason.startsWith(`Reuse`))?.descriptor;
+
+  // If `strategies` doesn't include `Strategy.REUSE` or there's no existing
+  // version in the project, try computing it from the base descriptor
+  if (!selected) {
+    let {range} = descriptor;
+
+    // If the range is a tag, we have to resolve it into a semver version
+    if (!semver.validRange(range)) {
+      const originalCandidates = await resolver.getCandidates(descriptor, new Map<DescriptorHash,Package>(), resolveOptions);
+      range = structUtils.parseRange(originalCandidates[0].reference).selector;
+    }
+
+    // Make a descriptor with the `^<major>` range
+    // Note: semver can coerce ranges into versions
+    selected = structUtils.makeDescriptor(request, suggestUtils.Modifier.CARET + semver.coerce(range)!.major);
+  }
```

Comment:
I think we could do without `suggestUtils` nor the strategies, since all the information we need are static. What do you think of something like this?

```ts
const coercedRange = `${suggestUtils.Modifier.CARET}${semver.coerce(descriptor.range)!.major}`;
const atTypesDescriptor = structUtils.makeDescriptor(structUtils.makeIdent(`types`, typesName), coercedRange);

const projectSuggestions = miscUtils.mapAndFind(project.workspaces, workspace => {
  const regularDependencyHash = workspace.manifest.dependencies.get(descriptor.identHash)?.descriptorHash;
  const devDependencyHash = workspace.manifest.devDependencies.get(descriptor.identHash)?.descriptorHash;

  // We only want workspaces that depend the exact same range as the original package
  if (regularDependencyHash !== descriptor.descriptorHash && devDependencyHash !== descriptor.descriptorHash)
    return miscUtils.mapAndFind.skip;

  const atTypesDependencies: Array<[AllDependencies, Descriptor]> = [];

  for (const type of Manifest.allDependencies) {
    const atTypesDependency = workspace.manifest[type].get(atTypesDescriptor.identHash);
    if (typeof atTypesDependency === `undefined`)
      continue;

    atTypesDependencies.push([type, atTypesDependency]);
  }

  // We only want workspaces that also depend on the appropriate @types package
  if (atTypesDependencies.length === 0)
    return miscUtils.mapAndFind.skip;

  return atTypesDependencies;
});

if (typeof projectSuggestions !== `undefined`) {
  for (const [dependencyType, atTypesDescriptor] of projectSuggestions) {
    workspace.manifest[dependencyType].set(atTypesDescriptor.identHash, atTypesDescriptor);
  }
} else {
  workspace.manifest[dependencyTarget].set(atTypesDescriptor.identHash, atTypesDescriptor);
}
```

Reply:
That's really smart!
The only thing remaining now is to write some tests. 🙂